# Revision - Level 2 - 24/04/2026
- Revision about `.concat()` and `.merge()`

In [2]:
import pandas as pd

# ⚔️ MISSÃO 1 — Basic Concat

👉 Combine two small datasets:

Create:
- df_online_orders (first 2 rows from orders)
- df_store_orders (last 2 rows from orders)

👉 Task:
- Concatenate them vertically

👉 Questions:
- What happens to the index?
- How to fix it?

In [4]:
df = pd.read_csv('orders.csv')
df

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150
2,103,1,1003,300
3,104,5,1002,120


In [5]:
df_online_orders = df.head(2)
df_online_orders

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150


In [6]:
df_store_orders = df.tail(2)
df_store_orders

,order_id,customer_id,product_id,amount
2,103,1,1003,300
3,104,5,1002,120


In [8]:
df_orders_concat = pd.concat([df_online_orders, df_store_orders], ignore_index=True)
df_orders_concat

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150
2,103,1,1003,300
3,104,5,1002,120


The indexes won't follow an order, the way to fix it is to use the argument `ignore_index` that way the index will follow the correct numerical order

# ⚔️ MISSÃO 2 — Horizontal Concat

Create:
- df_customer_names → only name column
- df_customer_cities → only city column

👉 Task:
- Concatenate horizontally (axis=1)

👉 Question:
- What happens if indexes don’t match?

In [9]:
df2 = pd.read_csv('customers.csv')
df2

,customer_id,name,city
0,1,Alice,New York
1,2,Bob,Chicago
2,3,Carol,Miami
3,4,Dave,Seattle


In [11]:
df_customer_names = df2[['name']]
df_customer_names

,name
0,Alice
1,Bob
2,Carol
3,Dave


In [20]:
df_customer_cities = df2[['city']]
new_row = pd.DataFrame({'city': 'Washigton D.C'}, index=[4])
df_customer_cities = pd.concat([df_customer_cities, new_row], ignore_index=True)
df_customer_cities

,city
0,New York
1,Chicago
2,Miami
3,Seattle
4,Washigton D.C


In [21]:
df_customers_name_city = pd.concat([df_customer_names, df_customer_cities], axis=1)
df_customers_name_city

,name,city
0,Alice,New York
1,Bob,Chicago
2,Carol,Miami
3,Dave,Seattle
4,NaN,Washigton D.C


When using `.concat()` if an index does not match, it creates a new row, with one of the values, with the dataframe that does not have a matched index, as NaN.

# ⚔️ MISSÃO 3 — Inner Join

👉 Merge:
- orders + customers

👉 Task:
- Use INNER JOIN on customer_id

👉 Questions:
- Which rows disappeared?
- Why?

In [24]:
df_orders = pd.read_csv('orders.csv')
df_orders

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150
2,103,1,1003,300
3,104,5,1002,120


In [25]:
df_customers = pd.read_csv('customers.csv')
df_customers

,customer_id,name,city
0,1,Alice,New York
1,2,Bob,Chicago
2,3,Carol,Miami
3,4,Dave,Seattle


In [26]:
df_customers_orders = pd.merge(df_orders, df_customers, on='customer_id', how='inner')
df_customers_orders

,order_id,customer_id,product_id,amount,name,city
0,101,1,1001,200,Alice,New York
1,102,2,1002,150,Bob,Chicago
2,103,1,1003,300,Alice,New York


from `orders.csv` the last row, index 3. and from `customers.csv` the rows 3 and 4, indexes 2 and 3,  disappeared, the motive is in the `inner` argument, when using this arugment we use an INNER JOIN just like in SQL, so only the `customer_id` that appeared in both data would be included in the merge.

# ⚔️ MISSÃO 4 — Left Join (IMPORTANT)

👉 Merge:
- orders + customers

👉 Task:
- Use LEFT JOIN

👉 Questions:
- Which rows have NaN?
- What does that mean in real life?

In [28]:
df_orders

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150
2,103,1,1003,300
3,104,5,1002,120


In [29]:
df_customers

,customer_id,name,city
0,1,Alice,New York
1,2,Bob,Chicago
2,3,Carol,Miami
3,4,Dave,Seattle


In [27]:
df_customers_orders = pd.merge(df_orders, df_customers, on='customer_id', how='left')
df_customers_orders

,order_id,customer_id,product_id,amount,name,city
0,101,1,1001,200,Alice,New York
1,102,2,1002,150,Bob,Chicago
2,103,1,1003,300,Alice,New York
3,104,5,1002,120,NaN,NaN


The NaN that appeared on the last row, happend because the `customers.csv` do not have a `customer_id` 5, so when we merged using `LEFT JOIN` it merged all the matching values from both DataFrames, and the values that only appeared on the LEFT DataFrame, in our case the `df_orders`  

# ⚔️ MISSÃO 5 — Right Join

👉 Merge:
- orders + customers

👉 Task:
- Use RIGHT JOIN

👉 Question:
- Which customers have no orders?

In [36]:
df_customers_orders = pd.merge(df_orders, df_customers, on='customer_id', how='right')
df_customers_orders

,order_id,customer_id,product_id,amount,name,city
0,101.0,1,1001.0,200.0,Alice,New York
1,103.0,1,1003.0,300.0,Alice,New York
2,102.0,2,1002.0,150.0,Bob,Chicago
3,NaN,3,NaN,NaN,Carol,Miami
4,NaN,4,NaN,NaN,Dave,Seattle


The customers 3 and 4 do not have orders, this happend similarly on mission 4, but this time the values that do not have a match came from the `customers.csv` DataFrame.

# ⚔️ MISSÃO 6 — Outer Join

👉 Merge:
- orders + customers

👉 Task:
- Use OUTER JOIN

👉 Question:
- What new information appears?

In [31]:
df_customers_orders = pd.merge(df_orders, df_customers, on='customer_id', how='outer')
df_customers_orders

,order_id,customer_id,product_id,amount,name,city
0,101.0,1,1001.0,200.0,Alice,New York
1,103.0,1,1003.0,300.0,Alice,New York
2,102.0,2,1002.0,150.0,Bob,Chicago
3,NaN,3,NaN,NaN,Carol,Miami
4,NaN,4,NaN,NaN,Dave,Seattle
5,104.0,5,1002.0,120.0,NaN,NaN


Now all the values from both DataFrames are merged together in this new DataFrame, 

# ⚔️ MISSÃO 7 — Multi Merge

👉 Merge:
- orders + customers + products

👉 Task:
1. First merge orders + customers
2. Then merge with products

👉 Goal:
Create a full dataset with:
- customer name
- product name
- category
- amount


In [38]:
df_orders = pd.read_csv('orders.csv')
df_orders

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150
2,103,1,1003,300
3,104,5,1002,120


In [39]:
df_customers = pd.read_csv('customers.csv')
df_customers

,customer_id,name,city
0,1,Alice,New York
1,2,Bob,Chicago
2,3,Carol,Miami
3,4,Dave,Seattle


In [40]:
df_products = pd.read_csv('products.csv')
df_products

,product_id,product_name,category
0,1001,Laptop,Electronics
1,1002,Shoes,Clothing
2,1003,Phone,Electronics
3,1004,Watch,Accessories


In [43]:
df_orders_customers = pd.merge(df_orders, df_customers, on='customer_id', how='outer')
df_orders_customers

,order_id,customer_id,product_id,amount,name,city
0,101.0,1,1001.0,200.0,Alice,New York
1,103.0,1,1003.0,300.0,Alice,New York
2,102.0,2,1002.0,150.0,Bob,Chicago
3,NaN,3,NaN,NaN,Carol,Miami
4,NaN,4,NaN,NaN,Dave,Seattle
5,104.0,5,1002.0,120.0,NaN,NaN


In [44]:
df_orders_customers_products = pd.merge(df_orders_customers, df_products, on='product_id', how='outer')
df_orders_customers_products

,order_id,customer_id,product_id,amount,name,city,product_name,category
0,101.0,1.0,1001.0,200.0,Alice,New York,Laptop,Electronics
1,102.0,2.0,1002.0,150.0,Bob,Chicago,Shoes,Clothing
2,104.0,5.0,1002.0,120.0,NaN,NaN,Shoes,Clothing
3,103.0,1.0,1003.0,300.0,Alice,New York,Phone,Electronics
4,NaN,NaN,1004.0,NaN,NaN,NaN,Watch,Accessories
5,NaN,3.0,NaN,NaN,Carol,Miami,NaN,NaN
6,NaN,4.0,NaN,NaN,Dave,Seattle,NaN,NaN


# ⚔️ MISSÃO 8 — Data Quality Check

👉 Questions:

- Are there orders with missing customers?
- Are there products never sold?

👉 Task:
- Identify both cases using merge logic

1. Yes, there is an order with no id and no customer
2. Yes, there are three products that were never sold

# ⚔️ MISSÃO 9 — Business Insight

👉 Analyze:

- Which category generates more revenue?
- Which customer spent the most?

👉 Deliver:
- Groupby + interpretation

In [46]:
df_orders_customers_products.groupby('category')['amount'].sum()

category
Accessories      0.0
Clothing       270.0
Electronics    500.0
Name: amount, dtype: float64

Eletronics generate the most revenue

In [47]:
df_orders_customers_products.groupby('customer_id')['amount'].sum()

customer_id
1.0    500.0
2.0    150.0
3.0      0.0
4.0      0.0
5.0    120.0
Name: amount, dtype: float64

The customer 1, spent the most

# ⚔️ MISSÃO 10 — HARD MODE 🧠

👉 Scenario:

A new dataset arrives:

## 🆕 new_orders
| order_id | customer_id | product_id | amount |
|----------|------------|------------|--------|
| 105      | 3          | 1004       | 220    |
| 106      | 6          | 1001       | 500    |

👉 Tasks:

1. Append to existing orders using concat
2. Merge everything again
3. Identify:
   - unknown customers
   - valid transactions
   - inconsistencies

In [59]:
df_orders = pd.read_csv('orders.csv')
df_orders

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150
2,103,1,1003,300
3,104,5,1002,120


In [60]:
df_new_orders = pd.read_csv('new_orders.csv')
df_new_orders

,order_id,customer_id,product_id,amount
0,105,3,1004,220
1,106,6,1001,500


In [61]:
df_customers = pd.read_csv('customers.csv')
df_customers

,customer_id,name,city
0,1,Alice,New York
1,2,Bob,Chicago
2,3,Carol,Miami
3,4,Dave,Seattle


In [62]:
df_products = pd.read_csv('products.csv')
df_products

,product_id,product_name,category
0,1001,Laptop,Electronics
1,1002,Shoes,Clothing
2,1003,Phone,Electronics
3,1004,Watch,Accessories


In [63]:
# First merge both order and new order
df_orders = pd.concat([df_orders, df_new_orders], ignore_index=True)
df_orders

,order_id,customer_id,product_id,amount
0,101,1,1001,200
1,102,2,1002,150
2,103,1,1003,300
3,104,5,1002,120
4,105,3,1004,220
5,106,6,1001,500


In [64]:
# Next merge both orders and customers
df_customers_orders = pd.merge(df_customers, df_orders, on='customer_id', how='outer')
df_customers_orders

,customer_id,name,city,order_id,product_id,amount
0,1,Alice,New York,101.0,1001.0,200.0
1,1,Alice,New York,103.0,1003.0,300.0
2,2,Bob,Chicago,102.0,1002.0,150.0
3,3,Carol,Miami,105.0,1004.0,220.0
4,4,Dave,Seattle,NaN,NaN,NaN
5,5,NaN,NaN,104.0,1002.0,120.0
6,6,NaN,NaN,106.0,1001.0,500.0


In [65]:
# Last merge with products
df_customers_orders_products = pd.merge(df_customers_orders, df_products, on='product_id', how='outer')
df_customers_orders_products

,customer_id,name,city,order_id,product_id,amount,product_name,category
0,1,Alice,New York,101.0,1001.0,200.0,Laptop,Electronics
1,6,NaN,NaN,106.0,1001.0,500.0,Laptop,Electronics
2,2,Bob,Chicago,102.0,1002.0,150.0,Shoes,Clothing
3,5,NaN,NaN,104.0,1002.0,120.0,Shoes,Clothing
4,1,Alice,New York,103.0,1003.0,300.0,Phone,Electronics
5,3,Carol,Miami,105.0,1004.0,220.0,Watch,Accessories
6,4,Dave,Seattle,NaN,NaN,NaN,NaN,NaN


1. Index 1 has an unkown customer
2. Dave does not have a valid transaction since all the products details are NaN